tests/unit/test_provider.py

```diff
@@ -22,382 +18,81 @@
-    )
-    harness.update_relation_data(
-        harness.charm.provider.client_relations[1].id, "new_application", {"chroot": "new_app"}
-    )
-
-    with patch("provider.generate_password", return_value="mellon"):
-        relations_config = harness.charm.provider.relations_config()
-
-    assert relations_config == {
-        "0": {
-            "username": "relation-0",
-            "password": "mellon",
-            "chroot": "/app",
-            "acl": "cdrwa",
-            "acls-added": "false",
-        },
-        str(new_rel_app_id): {
-            "username": f"relation-{new_rel_app_id}",
-            "password": "mellon",
-            "chroot": "/new_app",
-            "acl": "cdrwa",
-            "acls-added": "false",
-        },
-    }
-
-
-def test_build_acls(harness):
-    rel_id_new_app = harness.add_relation(REL_NAME, "new_application")
-    harness.update_relation_data(
-        harness.charm.provider.client_relations[0].id, "application", {"chroot": "app"}
-    )
-    harness.update_relation_data(
-        harness.charm.provider.client_relations[1].id,
-        "new_application",
-        {"chroot": "new_app", "chroot-acl": "rw"},
-    )
-
-    acls = harness.charm.provider.build_acls()
-
-    assert len(acls) == 2
-    assert sorted(acls.keys()) == ["/app", "/new_app"]
-    assert isinstance(acls["/app"], list)
-
-    new_app_acl = acls["/new_app"][0]
-
-    assert new_app_acl.acl_list == ["READ", "WRITE"]
-    assert new_app_acl.id.scheme == "sasl"
-    assert new_app_acl.id.id == f"relation-{rel_id_new_app}"
-
-
-def test_relations_config_values_for_key(harness):
-    rel_id_new_app = harness.add_relation(REL_NAME, "new_application")
-    harness.update_relation_data(
-        harness.charm.provider.client_relations[0].id, "application", {"chroot": "app"}
-    )
-    harness.update_relation_data(
-        harness.charm.provider.client_relations[1].id,
-        "new_application",
-        {"chroot": "new_app", "chroot-acl": "rw"},
-    )
-
-    config_values = harness.charm.provider.relations_config_values_for_key(key="username")
-
-    assert config_values == {f"relation-{rel_id_new_app}", "relation-0"}
-
-
-def test_is_child_of(harness):
-    chroot = "/gandalf/the/white"
-    chroots = {"/gandalf", "/saruman"}
-
-    assert harness.charm.provider._is_child_of(path=chroot, chroots=chroots)
-
-
-def test_is_child_of_not(harness):
-    chroot = "/the/one/ring"
-    chroots = {"/gandalf", "/saruman"}
+def test_client_relation_updated_defers_if_not_stable_leader(harness):
+    with (
+        patch("core.cluster.ClusterState.stable", new_callable=PropertyMock, return_value=False),
+        patch("ops.framework.EventBase.defer") as patched_defer,
+        patch("managers.quorum.QuorumManager.update_acls") as patched_acls,
+    ):
+        app_id = harness.add_relation(REL_NAME, "application")
+        harness.update_relation_data(app_id, "application", {"chroot": "balrog"})
```

Comment:
Being unable to match this much of tests re-shuffle...

Did the Provider tests module shrink so much because contents went to another module?

Reply:
Those tests relied on a now-removed `relation_config` object (a pseudo DPLibs before it existed), so are redundant.